internal/client/git.go

```diff
@@ -0,0 +1,190 @@
+	}); err != nil {
+		return fmt.Errorf("failed to setup local Brew repo: %w", err)
+	}
+
+	if err := os.WriteFile(filepath.Join(ctx.Config.Dist, repoName, path), content, 0o644); err != nil {
+		return fmt.Errorf("failed to write %s: %w", path, err)
+	}
+
+	log.WithField("repo", url).WithField("name", path).Info("pushing")
+	if err := runCmds(ctx, filepath.Join(ctx.Config.Dist, repoName), env, [][]string{
+		{"add", "-A", "."},
+		{"commit", "-m", msg},
+		{"push", "origin", "HEAD"},
+	}); err != nil {
+		return fmt.Errorf("failed to push %q (%q): %w", path, url, err)
+	}
+
+	return nil
+}
+
+func runCmds(ctx *context.Context, cwd string, env []string, cmds [][]string) error {
+	for _, cmd := range cmds {
+		args := append([]string{"-C", cwd}, cmd...)
+		if _, err := git.Clean(git.RunWithEnv(ctx, env, args...)); err != nil {
+			return fmt.Errorf("%q failed: %w", strings.Join(cmd, " "), err)
+		}
+	}
+	return nil
+}
+
+func keyPath(key string) (string, error) {
+	if key == "" {
+		return "", pipe.Skip("private_key is empty")
+	}
+
+	path := key
+	if _, err := ssh.ParsePrivateKey([]byte(key)); err == nil {
+		// if it can be parsed as a valid private key, we write it to a
+		// temp file and use that path on GIT_SSH_COMMAND.
+		f, err := os.CreateTemp("", "id_*")
+		if err != nil {
+			return "", fmt.Errorf("failed to store private key: %w", err)
+		}
+		defer f.Close()
+
+		// the key needs to EOF at an empty line, seems like github actions
+		// is somehow removing them.
+		if !strings.HasSuffix(key, "\n") {
+			key += "\n"
+		}
+
+		if _, err := io.WriteString(f, key); err != nil {
+			return "", fmt.Errorf("failed to store private key: %w", err)
+		}
+		if err := f.Close(); err != nil {
+			return "", fmt.Errorf("failed to store private key: %w", err)
+		}
+		path = f.Name()
+	}
+
+	if _, err := os.Stat(path); err != nil {
+		return "", fmt.Errorf("could not stat private_key: %w", err)
+	}
+
+	// in any case, ensure the key has the correct permissions.
+	if err := os.Chmod(path, 0o600); err != nil {
+		return "", fmt.Errorf("failed to ensure private_key permissions: %w", err)
+	}
+
+	return path, nil
+}
+
+func (gc *gitClient) CloseMilestone(ctx *context.Context, repo Repo, title string) (err error) {
+	return ErrNotImplemented
+}
+
+func (gc *gitClient) CreateRelease(ctx *context.Context, body string) (releaseID string, err error) {
+	return "", ErrNotImplemented
+}
+
+func (gc *gitClient) ReleaseURLTemplate(ctx *context.Context) (string, error) {
+	return "", ErrNotImplemented
+}
+
+func (gc *gitClient) Upload(ctx *context.Context, releaseID string, artifact *artifact.Artifact, file *os.File) (err error) {
+	return ErrNotImplemented
+}
+
+func (gc *gitClient) GetDefaultBranch(ctx *context.Context, repo Repo) (string, error) {
+	return "", ErrNotImplemented
+}
+
+func (gc *gitClient) Changelog(ctx *context.Context, repo Repo, prev, current string) (string, error) {
+	return "", ErrNotImplemented
+}
```

Comment:
TODO for me: segregate this interface, its too big and not all methods are needed everywhere its used...